engineering-plans/0004-subgraph-grafting.md

```diff
@@ -0,0 +1,170 @@
+# PLAN-0004: Subgraph Grafting
+
+<dl>
+  <dt>Author</dt>
+  <dd>David Lutterkort</dd>
+
+  <dt>Implements</dt>
+  <dd>No RFC</dd>
+
+  <dt>Engineering Plan pull request</dt>
+  <dd><a href="URL">URL</a></dd>
+
+  <dt>Date of submission</dt>
+  <dd>2020-03-17</dd>
+
+  <dt>Date of approval</dt>
+  <dd>YYYY-MM-DD</dd>
+
+  <dt>Approved by</dt>
+  <dd>First Person, Second Person</dd>
+</dl>
+
+## Contents
+
+<!-- toc -->
+
+## Summary
+
+This feature makes it possible to resume indexing of a failed subgraph by
+creating a new subgraph that uses the failed subgraph's data up to a block
+that the subgraph writer deems reliable, and continue indexing using the
+new subgraph's mappings. We call this operation *grafting* subgraphs.
+
+The main use case for this feature is to allow subgraph developers to move
+quickly past an error in a subgraph, especially during development. The
+overarching goal should always be to replace a subgraph that was grafted
+with one that was indexed from scratch as soon as that is reasonably
+possible.
+
+## Subgraph manifest changes
+
+The user-facing control for this feature will be a new optional
+`deployment` field in the subgraph's manifest. It can take one of three
+forms, and defaults to `create`:
+
+    ---
+    description: My really very good subgraph
+    # Option 1: (default) The deployment starts with no data in the database
+    deployment:
+      mode: create
+    # Option 2: (copy) The deployment starts with a copy of the data from
+    # the given subgraph at the given block
+    deployment:
+      mode: copy
+      subgraph: Qm...
+      block: 123456
+    # Option 3: (reuse) The deployment reuses the data of an existing subgraph
+    # in place. This mode implicitly deletes the source deployment, and any
+    # subgraph version using the source deployment will be switched to use the
+    # new deployment
+    deployment:
+      mode: reuse
+      subgraph: Qm...
+      block: 123456
+
+- For expediency, we will first implement `copy` and see if `reuse` is
+  actually needed; `reuse` is very invasive and can interact badly with
+  other features (e.g., sharing of subgraph deployments under several
+  names). The big advantage of `reuse` is that it does not require copying
+  data and will therefore be faster to initialize a subgraph, but it
+  remains to be seen how much of a factor that is in practice
+- For `copy` and `reuse` it is a deploy-time error if the source subgraph
+  does not exist.
+- For `copy` and `reuse`, the block up to which we use the source subgraph
+  must be final. It must therefore be at least `ETHEREUM_REORG_THRESHOLD`
+  blocks behind the chain head, and be on the main chain.
```

Comment:
This is a problem, because it implies a minimum downtime of `ETHEREUM_REORG_THRESHOLD` blocks.  On the hosted service (current reorg threshold: 200 blocks), this corresponds to roughly 50 minutes (at 15s block times).

Can't we require a block hash to continue from, identify what block number that corresponds to, fail if we can't, and otherwise copy the source subgraph based on that number? If this block were to be reorged later, the subgraph would handle that just like the original.

Reply:
After talking to @leoyvens, I realize that the situation is more complicated.

When grafting against block `N` but block `N` gets reorg'ed later, your grafting base block doesn't even exist anymore. Which is weird. Suddenly you have a different base block...?

In a sense, grafting demands that the base block is treated as final. Since we can't guarantee that, @leoyvens suggested failing whenever the grafting base block is reorg'ed, which I agree with.

This would allow us to use more recent blocks for grafting. The likelihood that a block mined 5 minutes ago is already very, very low, so in reality, we should very rarely see a grafting block being reorganized and causing such a failure.

Reply:
I agree that that is a pragmatic compromise between making people wait forever to be allowed to graft and risking undefined behavior. And that also gives a reason for storing the base block for a graft with the subgraph, as that needs to be checked now for each subgraph for each revert.

As an aside, for grafting, it's enough to deal with block numbers rather than hashes as they are unique in the context of any given subgraph.